kratos/utilities/assign_mpcs_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,376 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main author:     Sebastian Ares de Parga Regalado
+//
+
+// System includes
+#include <string>
+#include <iostream>
+
+// External includes
+#include "concurrentqueue/concurrentqueue.h"
+
+// Project includes
+#include "utilities/assign_mpcs_to_neighbours_utility.h"
+
+namespace Kratos 
+{
+///@addtogroup Kratos Core
+///@{
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @class AssignMPCsToNeighboursUtility
+ * @ingroup Kratos Core
+ * @brief Assing MPCs to Neighbouring Nodes
+ * @details This class provides a method to search for neighbouring nodes of one node
+ * within a given radius and assign a multipoint constraint (MPC) using radial basis functions.
+ * @author Sebastian Ares de Parga Regalado
+ */
+
+/// Default constructor.
+AssignMPCsToNeighboursUtility::AssignMPCsToNeighboursUtility(NodesContainerType& rStructureNodes){
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_model_part = rStructureNodes.GetContainer();
+    mpBins = Kratos::make_unique<NodeBinsType>(nodes_model_part.begin(), nodes_model_part.end());
+    mMaxNumberOfNodes = rStructureNodes.size();
+    KRATOS_CATCH(""); 
+} 
+
+// Search for the neighbouring nodes (in rStructureNodes) of each rNode on a given array of rNodes
+void AssignMPCsToNeighboursUtility::SearchNodesInRadiusForNodes(
+    NodesContainerType const& rNodes,
+    const double Radius,
+    VectorResultNodesContainerType& rResults,
+    const double MinNumOfNeighNodes)
+{
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_array     = const_cast<NodesContainerType::ContainerType&>(rNodes.GetContainer());
+
+    rResults.resize(nodes_array.size());
+
+    // Declare a counter variable outside the loop as std::atomic<int>
+    std::atomic<int> i(0);
+
+    block_for_each(nodes_array, [&](Node<3>::Pointer& pNode)
+    {
+        double localRadius = Radius;
+
+        IndexType it = i.fetch_add(1); // Atomically increment the counter and get the previous value
+
+        while (rResults[it].size()<MinNumOfNeighNodes){
+            rResults[it].clear();
+            SearchNodesInRadiusForNode(pNode, localRadius, rResults[it]);
+            localRadius += Radius;
+        }
+    });
+    KRATOS_CATCH("");
+}
+
+
+//Search for the neighbouring nodes (in rStructureNodes) of the given rNode
+void AssignMPCsToNeighboursUtility::SearchNodesInRadiusForNode(
+    NodeType::Pointer pNode,
```

Comment:
I'd pass a constant reference instead. Note that the `SearchInRadius` supports it.

Reply:
Do you find it necesarry, since this is a constant pointer. Isn't it okay like this?

Reply:
I don't think it is a constant pointer but nevermind, leave it like this.